docs/debug.md

```diff
@@ -104,8 +105,122 @@ Breakpoint 1, tee_entry_std (smc_args=0x7df6ff98 <stack_thread+8216>)
 (gdb)
 ```
 
-## 1.5. Use graphical frontends
-### 1.5.1 ddd
+## 1.5 Debugging TA
+Assumptions:
+1.	You already know how to setup `gdb` from the previous chapter;
+2.	Hello World TA will be used to debug as an example;
+3.	`pwd` is `…/optee/hello_world/ta`
+
+First, we need to find out the LMA (Load Memory Address) of Hello World TA.
+To do that run `objdump` utility (you might need to use platform specific one
+from the toolchain of the platform you are building for):
+```bash
+$ objdump -h 8aaaf200-2450-11e4-abe2-0002a5d5c51b.elf
+```
+
+The result will look something like this:
+```
+Sections:
+Idx   Name       Size      VMA               LMA               File off  Algn
+   0   .ta_head  00000020  0000000000000000  0000000000000000  00010000  2**3
+                 CONTENTS, ALLOC, LOAD, DATA
+   1   .text     0000c614  0000000000000020  0000000000000020  00010020  2**3
+                 CONTENTS, ALLOC, LOAD, READONLY, CODE
+   2   .rodata   00002382  000000000000c638  000000000000c638  0001c638  2**3
+```
+
+We are interested in the LMA (Load Memory Address) of `.text` segment. It’s the
+address where the elf will be loaded to. 
+
+Remember, TA is loaded dynamically at runtime by kernel driver `tee_supplicant`
+when CA is requesting access to TA. That’s why LMA by itself cannot be used to
+determine the actual address of the TA loaded into OP-TEE OS. But we will need
+this value later on.
+
+Now, you can run gdb and connect to QEMU session.
+
+```
+(gdb) optee
+SIGTRAP is used by the debugger.
+Are you sure you want to change it? (y or n) [answered Y; input not from
+terminal]
+0x00000000 in ?? ()
+```
+
+We have to set a breakpoint somewhere inside OP-TEE OS. For example:
+```
+(gdb) b tee_ta_invoke_command
+Breakpoint 1 at 0xe111880: file core/kernel/tee_ta_manager.c, line 598.
```

Comment:
I tested this on the qemu_virt setup. The `b TA_InvokeCommandEntryPoint` (line 204) fails because the TA is not yet mapped to mmu when `tee_ta_invoke_command` is reached. Hence gdb refuses to load a breakpoint on symbol `TA_InvokeCommandEntryPoint`.
I had to break from `user_ta_enter()` once `tee_ta_push_current_session(session)` is called. 
The best place I could find to break with TA mapped is `thread_enter_user_mode()` instead of `tee_ta_invoke_command()`.

Reply:
Yeah! That's a cool finding, coz' now you can break at any stage of TA invocation,  like: TA_CreateEntryPoint, TA_OpenSessionEntryPoint, TA_InvokeCommandEntryPoint.
I will edit that in. Thx.